### PR TITLE
feat: create edit employee page

### LINKED ISSUE #87: Create edit employee page
Refs: [Figma](https://www.figma.com/design/JvzZN6eHqoo2zgx6S6qbpT/EcoMap?node-id=1086-4359&t=mTXNi4vO6a8oJfgH-0)

### PR DESCRIPTION
**Pull request by goncalojdias**
Refs: closes #87 

## Summary

Create employee edit page.

### CODE REVIEW

File: web/src/routes/backOffice/employees/details/Employee.svelte [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	import { DateFormats } from "../../../../lib/constants/date";
+	import { getToastContext } from "../../../../lib/contexts/toast";
+	import { BackOfficeRoutes } from "../../../constants/routes";
+
+	/**
+	 * Employee ID.
+	 */
+	export let id: string;
+
+	const toast = getToastContext();
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/details/Employee.svelte]:
```suggestion
       /**
         * Toast context.
         */
	const toast = getToastContext();
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+					label={$t("employees.lastName")}
+					error={!!formErrorMessages.lastName}
+					helperText={formErrorMessages.lastName}
+				>
+					<Input
+						required
+						name="lastName"
+						value={employee?.lastName}
+						error={!!formErrorMessages.lastName}
+						placeholder={$t("employees.firstName.placeholder")}
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]:
```suggestion
						placeholder={$t("employees.lastName.placeholder")}
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+						minLength={formFieldsLengths.firstName.min}
+						maxLength={formFieldsLengths.firstName.max}
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]:
```suggestion
						minLength={formFieldsLengths.lastName.min}
						maxLength={formFieldsLengths.lastName.max}
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+						error={!!formErrorMessages.username}
+						placeholder={$t("employees.username.placeholder")}
+						minLength={formFieldsLengths.username.min}
+						maxLength={formFieldsLengths.username.max}
+					/>
+				</FormControl>
+
+				<!-- Birthdate -->
+				<FormControl
+					label={$t("employees.birthdate")}
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: Just for consistency purposes, change every place that contains `birthdate` to `dateOfBirth`

File: web/src/locales/en.json [hunk 5]
```diff
@@ -110,5 +112,23 @@
...
+	"employees.delete.success": "Employee deleted successfully",
+	"employees.delete.conflict.title": "Conflict",
+	"employees.delete.conflict.description": "This employee cannot be deleted because it's associated with a report or a route.",
+	"employees.update.success": "Employee updated successfully",
+	"employees.username.placeholder": "Enter username",
+	"employees.firstName.placeholder": "Enter first name",
+	"employees.lastName.placeholder": "Enter last name",
+	"employees.birthdate.placeholder": "Enter birthdate",
+	"employees.phone.placeholder": "Enter phone number",
+	"employees.scheduleStart.placeholder": "Enter Schedule start",
```
Comment (joaotomaspinheiro) [inline web/src/locales/en.json]:
```suggestion
	"employees.scheduleStart.placeholder": "Enter start time of schedule",
```

File: web/src/locales/en.json [hunk 6]
```diff
@@ -110,5 +112,23 @@
...
+	"employees.scheduleEnd.placeholder": "Enter Schedule end"
```
Comment (joaotomaspinheiro) [inline web/src/locales/en.json]:
```suggestion
	"employees.scheduleEnd.placeholder": "Enter end time of schedule"
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+			});
+		} else {
+			formErrorMessages.firstName = "";
+		}
+
+		// LastName Validation.
+		if (lastNameValidity.valueMissing) {
+			formErrorMessages.lastName = $t("error.valueMissing");
+		} else if (lastNameValidity.patternMismatch) {
+			formErrorMessages.lastName = $t("error.patternMismatch");
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: I believe that there isn't a pattern mismatch error for the last name field.

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+			});
+		} else {
+			formErrorMessages.lastName = "";
+		}
+
+		// DateOfBirth Validation.
+		if (dateOfBirthValidity.valueMissing) {
+			formErrorMessages.dateOfBirth = $t("error.valueMissing");
+		} else if (dateOfBirthValidity.patternMismatch) {
+			formErrorMessages.dateOfBirth = $t("error.patternMismatch");
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: Is there a possibility of a incorrect pattern for dates inputs given that they are not free-form fields?

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+		} else {
+			formErrorMessages.dateOfBirth = "";
+		}
+
+		// PhoneNumber Validation.
+		if (phoneNumberValidity.valueMissing) {
+			formErrorMessages.phoneNumber = $t("error.valueMissing");
+		} else if (phoneNumberValidity.patternMismatch) {
+			formErrorMessages.phoneNumber = $t("error.patternMismatch");
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte:194]: The phone number doesn't have a pattern specified

File: web/src/routes/backOffice/employees/details/Employee.svelte [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+		if (res.error) {
+			throw new Error("Failed to retrieve employee details");
+		}
+		return res.data;
+	}
+
+	/**
+	 * Deletes the employee displayed on the page.
+	 */
+	async function deleteEmployee() {
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/details/Employee.svelte:43]: The deletion feature should be in a different pull request.

**Comment by joaotomaspinheiro** [comment]
Since this PR also includes the employee details page and the edit and delete feature, update the PR title and include those issues in the PR description with the keyword "closes".

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte (hunk 8 above)
Comment (goncalojdias) [reply web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: don`t think so

Comment (goncalojdias) [reply web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: I think the same can be said for scheduleStart and scheduleEnd.

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte (hunk 9 above)
Comment (goncalojdias) [reply web/src/routes/backOffice/employees/components/EmployeeForm.svelte:194]: Does it make sense to add +351 to the pattern? Or country codes in general ?

Comment (Goncalo-Marques) [reply web/src/routes/backOffice/employees/components/EmployeeForm.svelte:194]: For now, please don't use a country code in the pattern.
The "+" is a special char in the URL and it would break the list filter when providing a phone number with "+".
This happens because the "+" is replaced with a " ".

File: web/src/locales/en.json [hunk 11]
```diff
@@ -124,5 +124,13 @@
 	"employees.notFound.description": "The employee you are looking for does not exist.",
 	"employees.delete.success": "Employee deleted successfully",
 	"employees.delete.conflict.title": "Conflict",
-	"employees.delete.conflict.description": "This employee cannot be deleted because it's associated with a report or a route."
+	"employees.delete.conflict.description": "This employee cannot be deleted because it's associated with a report or a route.",
```
Comment (Goncalo-Marques) [inline web/src/locales/en.json]: This was updated, now the report can not generate any conflicts:
```suggestion
	"employees.delete.conflict.description": "This employee cannot be deleted because it's associated with a route.",
```

File: web/src/locales/pt.json [hunk 12]
```diff
@@ -124,5 +124,13 @@
 	"employees.notFound.description": "O colaborador que procura não existe.",
 	"employees.delete.success": "Colaborador eliminado com sucesso",
 	"employees.delete.conflict.title": "Conflito",
-	"employees.delete.conflict.description": "Este colaborador não pode ser eliminado porque está associado a um reporte ou a uma rota."
+	"employees.delete.conflict.description": "Este colaborador não pode ser eliminado porque está associado a um reporte ou a uma rota.",
```
Comment (Goncalo-Marques) [inline web/src/locales/pt.json]:
```suggestion
	"employees.delete.conflict.description": "Este colaborador não pode ser eliminado porque está associado a uma rota.",
```

File: web/src/locales/pt.json [hunk 13]
```diff
@@ -124,5 +124,13 @@
...
+	"employees.update.success": "Colaborador atualizado com sucesso",
+	"employees.username.placeholder": "Introduza o nome de utilizador",
+	"employees.firstName.placeholder": "Introduza o primeiro nome",
+	"employees.lastName.placeholder": "Introduza o último nome",
+	"employees.dateOfBirth.placeholder": "Introduza a data de nascimento",
+	"employees.phone.placeholder": "Introduza número de telefone",
```
Comment (Goncalo-Marques) [inline web/src/locales/pt.json]: For consistency:
```suggestion
	"employees.phone.placeholder": "Introduza o número de telefone",
```

File: web/src/locales/schema.json [hunk 14]
```diff
@@ -375,6 +375,30 @@
...
+			"type": "string"
+		},
+		"employees.phone.placeholder": {
+			"type": "string"
+		},
+		"employees.scheduleStart.placeholder": {
+			"type": "string"
+		},
+		"employees.scheduleEnd.placeholder": {
+			"type": "string"
```
Comment (Goncalo-Marques) [inline web/src/locales/schema.json]: Are these fields even visible?
Unless these fields are going to be visible in the create page 🤔

File: web/src/locales/schema.json [hunk 15]
```diff
@@ -375,6 +375,30 @@
...
+		"employees.username.placeholder": {
+			"type": "string"
+		},
+		"employees.firstName.placeholder": {
+			"type": "string"
+		},
+		"employees.lastName.placeholder": {
+			"type": "string"
+		},
+		"employees.dateOfBirth.placeholder": {
```
Comment (Goncalo-Marques) [inline web/src/locales/schema.json]: Same question for this `dateOfBirth` field.

File: web/src/lib/components/Input.svelte [hunk 16]
```diff
@@ -31,6 +31,12 @@
 	 */
 	export let id: HTMLInputAttributes["id"] = null;
 
+	/**
+	 * Defines the field legal number intervals. Can be used to force seconds field in time input.
+	 * @default null
+	 */
+	export let step: HTMLInputAttributes["step"] = null;
```
Comment (Goncalo-Marques) [inline web/src/lib/components/Input.svelte]: I know that in the Swagger spec, time is defined with seconds included.
However, this was only done to be consistent with the database data type.
Visually, in the front end, I don't think the seconds are important for the schedule.

**Comment by Goncalo-Marques** [comment]
Please consider the following comment in this PR: https://github.com/Goncalo-Marques/ecomap/pull/213#discussion_r1615175006

File: web/src/locales/en.json [hunk 17]
```diff
@@ -139,5 +139,10 @@
 	"employees.updatePassword.error.incorrectPassword.description": "The password entered does not match your current password.",
 	"employees.error.passwordMismatch": "The password doesn't match the new password",
 	"employees.passwordConstraints": "Password must contain at least 14 characters, including one regular character, one special character, and one number",
-	"confirm": "Confirm"
+	"confirm": "Confirm",
+	"employees.update.success": "Employee updated successfully",
+	"employees.username.placeholder": "Enter username",
+	"employees.firstName.placeholder": "Enter first name",
+	"employees.lastName.placeholder": "Enter last name",
+	"employees.phone.placeholder": "Enter phone number"
```
Comment (Goncalo-Marques) [inline web/src/locales/en.json]: This seems to be in the wrong place, after the merge.
Shouldn't it be place after `employees.delete.conflict.description`?

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 18]
```diff
@@ -0,0 +1,488 @@
...
+		);
+	}
+
+	/**
+	 * Handles the submit event of the form.
+	 * @param e Submit event.
+	 */
+	function handleSubmit(e: SubmitEvent) {
+		const form = e.currentTarget as HTMLFormElement;
+		const formData = new FormData(form);
```
Comment (Goncalo-Marques) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: To keep consistency with one of the latest changes João made to the forms:
```suggestion
	function handleSubmit(
		e: Event & { currentTarget: EventTarget & HTMLFormElement },
	) {
		const form = e.currentTarget;
		const formData = new FormData(form);
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 19]
```diff
@@ -0,0 +1,490 @@
...
+			});
+		} else if (usernameValidity.tooLong) {
+			formErrorMessages.username = $t("error.tooLong", {
+				maxLength: formFieldsLengths.username.max,
+			});
+		} else {
+			formErrorMessages.username = "";
+		}
+
+		//FirstName Validation.
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]:
```suggestion
		// FirstName Validation.
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 20]
```diff
@@ -0,0 +1,490 @@
...
+
+	/**
+	 * Validates form and sets error messages on the form fields if they contain errors.
+	 * @param usernameValidity
+	 * @param firstNameValidity
+	 * @param lastNameValidity
+	 * @param dateOfBirthValidity
+	 * @param phoneNumberValidity
+	 * @param locationValidity
+	 * @param coordinate
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: Add a description to these parameters

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 21]
```diff
@@ -0,0 +1,490 @@
...
+					label={$t("employees.scheduleStart")}
+					error={!!formErrorMessages.scheduleStart}
+					helperText={formErrorMessages.scheduleStart}
+				>
+					<Input
+						required
+						name="scheduleStart"
+						value={employee?.scheduleStart
+							? formatTime24H(employee.scheduleStart)
+							: ""}
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]:
```suggestion
						value={employee
							? formatTime24H(employee.scheduleStart)
							: ""}
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 22]
```diff
@@ -0,0 +1,490 @@
...
+					label={$t("employees.scheduleEnd")}
+					error={!!formErrorMessages.scheduleEnd}
+					helperText={formErrorMessages.scheduleEnd}
+				>
+					<Input
+						required
+						name="scheduleEnd"
+						value={employee?.scheduleEnd
+							? formatTime24H(employee.scheduleEnd)
+							: ""}
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]:
```suggestion
						value={employee
							? formatTime24H(employee.scheduleEnd)
							: ""}
```

File: web/src/routes/backOffice/employees/edit/EditEmployee.svelte [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+	/**
+	 * Updates employee with given username, firstName, lastName, dateOfBirth, phoneNumber, location, scheduleStart and scheduleEnd.
+	 * @param username
+	 * @param firstName
+	 * @param lastName
+	 * @param dateOfBirth
+	 * @param phoneNumber
+	 * @param location
+	 * @param scheduleStart
+	 * @param scheduleEnd
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/edit/EditEmployee.svelte]: Add documentation to these parameters

File: web/src/routes/backOffice/employees/edit/EditEmployee.svelte [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+		username: string,
+		firstName: string,
+		lastName: string,
+		dateOfBirth: string,
+		phoneNumber: string,
+		location: GeoJSONFeaturePoint,
+		scheduleStart: string,
+		scheduleEnd: string,
+	) {
+		// Adding seconds to times. Necessary because api receives times with seconds.
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/edit/EditEmployee.svelte]:
```suggestion
		// Adding seconds to times. Necessary because API receives times with seconds.
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+						name="username"
+						value={employee?.username}
+						error={!!formErrorMessages.username}
+						placeholder={$t("employees.username.placeholder")}
+						minLength={formFieldsLengths.username.min}
+						maxLength={formFieldsLengths.username.max}
+					/>
+				</FormControl>
+
+				<!-- dateOfBirth -->
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte]: Just for consistency with the other comments.
```suggestion
				<!-- DateOfBirth -->
```

File: web/src/routes/backOffice/employees/components/EmployeeForm.svelte [hunk 26]
```diff
@@ -0,0 +1,488 @@
...
+				<FormControl
+					label={$t("employees.dateOfBirth")}
+					error={!!formErrorMessages.dateOfBirth}
+					helperText={formErrorMessages.dateOfBirth}
+				>
+					<Input
+						required
+						name="dateOfBirth"
```
Comment (joaotomaspinheiro) [inline web/src/routes/backOffice/employees/components/EmployeeForm.svelte:406]: Add a validation to prevent the date of birth from being in the future.

**Review by Goncalo-Marques — APPROVED** [review]

**Review by joaotomaspinheiro — APPROVED** [review]
